datumaro/plugins/camvid_format.py

```diff
@@ -0,0 +1,248 @@
+    @classmethod
+    def find_sources(cls, path):
+        subset_paths = [p for p in glob(osp.join(path, '*.txt'))
+            if osp.basename(p) != CamvidPath.LABELMAP_FILE]
+        sources = []
+        for subset_path in subset_paths:
+            sources += cls._find_sources_recursive(
+                subset_path, '.txt', 'camvid')
+        return sources
+
+
+LabelmapType = Enum('LabelmapType', ['camvid', 'source'])
+
+class CamvidConverter(Converter):
+    DEFAULT_IMAGE_EXT = '.png'
+
+    def __init__(self, extractor, save_dir,
+            apply_colormap=True, label_map=None, **kwargs):
+        super().__init__(extractor, save_dir, **kwargs)
+
+        self._apply_colormap = apply_colormap
+
+        if label_map is None:
+            label_map = LabelmapType.source.name
+        self._load_categories(label_map)
+
+    def apply(self):
+        subset_dir = self._save_dir
+        os.makedirs(subset_dir, exist_ok=True)
+
+        for subset_name, subset in self._extractor.subsets().items():
+            segm_list = []
+            for item in subset:
+                masks = []
+                for a in item.annotations:
+                    if a.type == AnnotationType.mask:
+                        masks.append(a)
+
+                if masks:
+                    compiled_mask = CompiledMask.from_instance_masks(masks)
+
+                    self.save_segm(
+                        osp.join(subset_dir, subset_name + CamvidPath.SEGM_DIR, item.id + CamvidPath.IMAGE_EXT),
+                        compiled_mask.class_mask)
+
+                    self._save_image(item,
+                        osp.join(subset_dir, subset_name, item.id + CamvidPath.IMAGE_EXT))
+
+                    segm_list.append(item.id)
+
+                self.save_segm_lists(subset_name, segm_list)
+        self.save_label_map()
+
+    def save_segm(self, path, mask, colormap=None):
+        if self._apply_colormap:
+            if colormap is None:
+                colormap = self._categories[AnnotationType.mask].colormap
+            mask = paint_mask(mask, colormap)
+        save_image(path, mask, create_dir=True)
+
+    def save_segm_lists(self, subset_name, segm_list):
+        if not segm_list:
+            return
+
+        ann_file = osp.join(self._save_dir, subset_name + '.txt')
+        with open(ann_file, 'w') as f:
+            for item in segm_list:
+                f.write('/%s/%s /%s/%s\n' % (subset_name, item + CamvidPath.IMAGE_EXT,
+                    subset_name + CamvidPath.SEGM_DIR, item + CamvidPath.IMAGE_EXT))
+
+    def save_label_map(self):
+        path = osp.join(self._save_dir, CamvidPath.LABELMAP_FILE)
+        write_label_map(path, self._label_map)
+
+    def _load_categories(self, label_map_source):
+        if label_map_source == LabelmapType.camvid.name:
+            # use the default Camvid colormap
+            label_map = CamvidLabelMap
+
+        elif label_map_source == LabelmapType.source.name and \
```

Comment:
We should also cover the case when `source`, but it does not have `MaskCategories`. It is the most used case - for example, it would be used after conversion of polygons to masks in a dataset. We also need to make sure `0,0,0` exists, and has index 0, otherwise masks will be wrong. Check VOC code for example. Maybe, it is worth trying to extract the common code fragment.

Reply:
How are we supposed to generate colormap if we don't have MaskCategories? Can we do the same as in VOC?

Reply:
Yes, let's use the same colormap generation algorithm.